### PR TITLE
Upgrade to ring 0.17

### PR DESCRIPTION
**Pull request by djc**
Question: do we want to backport this to rustls 0.21? Would enable usage of rustls a on bunch more targets.

### CODE REVIEW

**Comment by djc** [comment]
@aochagavia the instruction count differences seem pretty worrying. Could you run time-based tests on your fancy new server for this PR?

File: connect-tests/Cargo.toml
Comment (cpu) [inline connect-tests/Cargo.toml:5]: Why are these `rust-version` directives falling away here, maybe a candidate for a separate commit? The [workspace docs](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-package-table) read to me like we need a `[workspace.package]` in the top-level `Cargo.toml` for this to be inherited, is that incorrect?

File: fuzz/Cargo.toml
Comment (cpu) [inline fuzz/Cargo.toml:13]: This is another change I'm having trouble matching up to the *ring* update. Could you split it off?

**Review by cpu — COMMENTED** [review]
Generally looks good, just a couple nits/comments

**Comment by ctz** [comment]
> I could spend some time tomorrow to create a bunch of ad-hoc criterion benchmarks

I actually have some of these on the aws-lc-rs branch -- will extract them to merge separately.

**Comment by ctz** [comment]
> > I could spend some time tomorrow to create a bunch of ad-hoc criterion benchmarks
> 
> I actually have some of these on the aws-lc-rs branch -- will extract them to merge separately.

That is #1511 -- local runs after merging that onto this gives:

```
 name                                                             main.txt ns/iter  ring17.txt ns/iter  diff ns/iter   diff %  speedup 
 bench_ewouldblock                                                11                11                             0    0.00%   x 1.00 
 crypto::ring::kx::benchmarks::bench_ecdh_p256                    52,784            51,970                      -814   -1.54%   x 1.02 
 crypto::ring::kx::benchmarks::bench_ecdh_p384                    769,208           765,313                   -3,895   -0.51%   x 1.01 
 crypto::ring::kx::benchmarks::bench_x25519                       54,136            75,106                    20,970   38.74%   x 0.72 ⬅
 crypto::ring::sign::benchmarks::bench_ecdsa_p256_sha256          16,739            16,540                      -199   -1.19%   x 1.01 
 crypto::ring::sign::benchmarks::bench_ecdsa_p384_sha384          415,218           411,350                   -3,868   -0.93%   x 1.01 
 crypto::ring::sign::benchmarks::bench_eddsa                      15,920            41,516                    25,596  160.78%   x 0.38 ⬅
 crypto::ring::sign::benchmarks::bench_load_and_validate_eddsa    15,033            40,857                    25,824  171.78%   x 0.37 ⬅
 crypto::ring::sign::benchmarks::bench_load_and_validate_p256     11,426            11,321                      -105   -0.92%   x 1.01 
 crypto::ring::sign::benchmarks::bench_load_and_validate_p384     385,406           384,189                   -1,217   -0.32%   x 1.00 
 crypto::ring::sign::benchmarks::bench_load_and_validate_rsa2048  21,941            21,666                      -275   -1.25%   x 1.01 
 crypto::ring::sign::benchmarks::bench_load_and_validate_rsa4096  80,350            77,442                    -2,908   -3.62%   x 1.04 
 crypto::ring::sign::benchmarks::bench_rsa2048_pkcs1_sha256       500,895           502,346                    1,451    0.29%   x 1.00 
 crypto::ring::sign::benchmarks::bench_rsa2048_pss_sha256         507,693           503,370                   -4,323   -0.85%   x 1.01 
 tls12::prf::benchmarks::bench_sha256                             2,436             2,370                        -66   -2.71%   x 1.03 
 tls13::key_schedule::benchmarks::bench_sha256                    10,269            10,030                      -239   -2.33%   x 1.02 
```

**Comment by djc** [comment]
@briansmith are these performance changes to be expected?

**Comment by briansmith** [comment]
It looks like it is just Curve25519? The faster Curve25519 will come in *ring* 0.17.1. I will ping you with the PR when it is ready.

**Comment by briansmith** [comment]
@ctz what is the exact hardware you are running on?

FWIW, I do think the Curve25519 regressions would be expected although I thought they would be smaller than that. There is a better-optimized Curve25519 that I brought into *ring* 0.17.0 as part of the BoringSSL merge but didn't enable because I wanted to do some additional review/testing.

In the case of RSA, it looks like the benchmarks are noisy since we wouldn't expect PSS vs PKCS#1 to make a difference. If you could clean up the noise it would be great. BoringSSL did make a change that might have negatively affected performance slightly, and my analogous change might have affected performance slightly more than that. If so, it can be fixed.

Similarly, some other algorithms will also get faster because I brought in new faster code, but didn't hook it up. However, unlike Curve25519, there weren't any major changes that would have regressed performance, at least intentionally, IIRC.

**Comment by briansmith** [comment]
I would suggest waiting a bit before doing this, until more of the perf work in 0.17.1 is done, so that we can understand how that work affects the build requirements. It might require a newer GCC and/or clang than 0.17.0 requires right now, and I'd like to get that all clarified.

**Comment by briansmith** [comment]
@ctz https://github.com/briansmith/ring/pull/1688 is the PR for adding the new optimized Curve25519.

**Comment by ctz** [comment]
> @ctz what is the exact hardware you are running on?

Intel Skylake -- specifically i7-7820X.

File: README.md
Comment (ctz) [inline README.md:89]: edit this also in src/lib.rs

Comment (djc) [reply README.md:89]: (It was actually there, just in the other commit -- fixed that up now.)

File: connect-tests/Cargo.toml
Comment (djc) [reply connect-tests/Cargo.toml:5]: Pulled the MSRV changes out into #1516.

**Comment by djc** [comment]
Should we hold off on merging this until the performance regresssion is dealt with? Should we hold off on releasing a 0.21.x release with this until the performance regression is dealt with? (I'm thinking maybe not and yes.)

File: fuzz/Cargo.toml
Comment (cpu) [reply fuzz/Cargo.toml:13]: (I still think this cleanup is unrelated to the ring update and would be better in a separate commit)

**Comment by briansmith** [comment]
using `TARGET_CC=clang cargo bench x25519` with clang-16 and the latest cargo/rust, I get significantly better performance with https://github.com/briansmith/ring/pull/1688 according to the benchmarks at https://github.com/briansmith/ring/pull/1694.

**Comment by ctz** [comment]
> Should we hold off on merging this until the performance regresssion is dealt with? Should we hold off on releasing a 0.21.x release with this until the performance regression is dealt with? (I'm thinking maybe not and yes.)

I'd agree with that assessment.

This PR probably also depends on (or could incorporate) taking rustls-webpki==0.102.0-alpha.4 -- see eg fb8ee7578d2ad85d1c40f0bf07368efb0fb2fdb0

**Comment by djc** [comment]
Hmm, I think I disagree. Clearly there's a lot of excitement about ring 0.17 (given the amount of issues opened and the number of issues/PRs linking to this PR), probably because of the massively increased portability? The point of not having ring appear in the public API means that we can bump ring without this being externally visible. rustls 0.22 is probably at least a week from being released and even when we release it, it will take the ecosystem a while to update dependencies -- that means that, for the time being, downstreams might be likely to pull two versions of ring into their dependency graphs when there's really no good reason to do that. So I think we *should* have ring 0.17 backports for webpki 0.101 and rustls 0.21 and release them as soon as we think it's ready, but maybe after the worst of the performance regressions has been addressed.

Or do you see serious risks for rustls 0.21 from our upgrade to ring 0.17 (I don't really). Plus, once we release rustls 0.22 we're on the hook to provide rustls 0.21 maintenance for a year longer -- that can be quite limited but it might still be nice if they are both on the latest ring release.

**Comment by ctz** [comment]
> The point of not having ring appear in the public API means that we can bump ring without this being externally visible.

I'm not sure we have anything that validates we've achieved this though. I just went to check 0.21 and `rustls::cipher_suite::TLS13_AES_256_GCM_SHA384.hash_algorithm()` is pub and leaks a *ring* type. Are there downstream callers of that? Probably not. If there is maybe they will accept "we wrote that `pub` before `pub(crate)` existed, and it should've been `pub(crate)` all along".

I guess it would be interesting if `cargo-semver-checks` catches that detail if we do the backport.

> Or do you see serious risks for rustls 0.21 from our upgrade to ring 0.17 (I don't really).

I don't see this as about risk, more about the extent to which previous release branches are active vs just maintained.

**Comment by djc** [comment]
> > The point of not having ring appear in the public API means that we can bump ring without this being externally visible.
> 
> I'm not sure we have anything that validates we've achieved this though. I just went to check 0.21 and `rustls::cipher_suite::TLS13_AES_256_GCM_SHA384.hash_algorithm()` is pub and leaks a _ring_ type. Are there downstream callers of that? Probably not. If there is maybe they will accept "we wrote that `pub` before `pub(crate)` existed, and it should've been `pub(crate)` all along".

Hmm, that's fair. Well, if we expose that probably one reason to not backport to 0.21.

> I guess it would be interesting if `cargo-semver-checks` catches that detail if we do the backport.

That would be a really interesting test -- I suppose it would?

> > Or do you see serious risks for rustls 0.21 from our upgrade to ring 0.17 (I don't really).
> 
> I don't see this as about risk, more about the extent to which previous release branches are active vs just maintained.

That's fair. I would be inclined to say that the 0.21 branch should definitely be perceived as "active" since it's what ~everyone is using right now. That doesn't mean that we backport every feature and refactoring, but for something like this which (a) should be (mostly) semver-compatible and (b) makes downstream users` life better with broader compatibility and (c) low risk, I like the cost-benefit of backporting.

It's also really nice to stick with up-to-date dependencies to avoid having duplicates in dependency graphs...

File: fuzz/Cargo.toml
Comment (ctz) [reply fuzz/Cargo.toml:13]: taken the liberty of splitting this into #1523

Comment (djc) [reply fuzz/Cargo.toml:13]: Thanks!

**Comment by cpu** [comment]
>  I like the cost-benefit of backporting.

fwiw, I agree. I don't think it's something we should necessarily do for every dependency but I think for this case it makes sense.

**Review by cpu — APPROVED** [review]

**Review by ctz — APPROVED** [review]

**Comment by cpu** [comment]
Started a 0.21.x backport in https://github.com/rustls/rustls/pull/1525

**Comment by briansmith** [comment]
0.17.1 was released with the performance improvement to Ed25519.

**Comment by briansmith** [comment]
More of the performance optimizations added in recent BoringSSL changes will appear in upcoming 0.17.x releases. We are trying a new QA process for enabling these.

**Comment by djc** [comment]
I bumped the lockfile to 0.17.2 in #1528, but it didn't seem to help much with the performance regressions seen here.

**Comment by briansmith** [comment]
Here are the latest benchmarks: https://github.com/briansmith/ring/pull/1688#issuecomment-1754061489. The fastest combination is *ring* 0.17 using Clang 17.

**Pull request merged**
